s3fs/utils.py

```diff
@@ -10,6 +15,100 @@ def ignoring(*exceptions):
         pass
 
 
+try:
+    from contextlib import AsyncExitStack
+except ImportError:
+    # Since AsyncExitStack is not available for 3.6<=
+    # we'll create a simple implementation that imitates
+    # the basic functionality.
+    class AsyncExitStack:
+        def __init__(self):
+            self.contexts = []
+
+        async def enter_async_context(self, context):
+            self.contexts.append(context)
+            return await context.__aenter__()
+
+        async def aclose(self, *args):
+            args = args or (None, None, None)
+            for context in self.contexts:
+                await context.__aexit__(*args)
+
+        async def __aenter__(self):
+            return self
+
+        async def __aexit__(self, *args):
+            await self.aclose(*args)
+
+
+class S3BucketRegionCache:
+    # See https://github.com/aio-libs/aiobotocore/issues/866
+    # for details.
+
+    def __init__(self, session, **client_kwargs):
+        self._session = session
+        self._stack = AsyncExitStack()
+        self._client = None
+        self._client_kwargs = client_kwargs
+        self._buckets = {}
+        self._regions = {}
+
+    async def get_bucket_client(self, bucket_name=None):
+        # TODO: raise error when client_kwargs has region
```

Comment:
This one is important!
If you specify a region and bucket is in a different region, what happens?

Reply:
We should decide to either ignore or raise an error when one of these options are specified: `region`, `endpoint_url` and `use_accelerated_endpoint` (or something like that, don't recall now).

Reply:
I would assume that if the user specifies, we should trust that they know what they are doing. This is already possible now, so we would be breaking someone's workflow otherwise.

Reply:
> This is already possible now, so we would be breaking someone's workflow otherwise.

I'm asking for the case where user both specify `cache_regions` and `region` at the same time. Should we just ignore `cache_regions`, or raise like `ValueError("You can't use both region and cache_regions options")`. I'd say that we could simply ignore `cache_regions`, but curious what you'd think?

Reply:
OK, I see what you mean. I would normally say that this is a ValueError for kwargs that don't make sense together. However, I expect caching to  become default one release later, so I think we can imply cache_regions=False so long as the documentation says so.

In the implementation before this PR, what happens if you specify a region and then access a bucket in another region?